### PR TITLE
add dob type

### PR DESCRIPTION
**Pull request by jgrayson676**
## Description of the change

Adds a new `dob` type and fixes a bug with the tf cert url tests.

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change
- [ ] Technical Debt
- [ ] Documentation

## Related tickets

https://app.shortcut.com/active-prospect/story/40321/user-maps-dob-components

### Development and Testing

- [ ]  Lint rules pass locally.
- [ ]  The code changed/added as part of this pull request has been covered with tests, or this PR does not alter production code.
- [ ]  All tests related to the changed code pass in development, or tests are not applicable.

### Code Review

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached.
- [ ]  At least two engineers have been added as "Reviewers" on the pull request.
- [ ]  Changes have been reviewed by at least two other engineers who did not write the code.
- [ ]  This branch has been rebased off master to be current.

### Tracking 
- [ ]  Issue from Shortcut has a link to this pull request.
- [ ]  This PR has a link to the issue in Shortcut.

### QA
- [ ]  This branch has been deployed to staging and tested.

### CODE REVIEW

File: lib/types/dob.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+const components = [
+  { name: 'raw', type: 'string', description: 'The full date, only accessible to Integrations' },
+  { name: 'age', type: 'number', description: 'The age of the consumer, calculated as of today' },
+  { name: 'year', type: 'number', description: 'The year from the DOB' }
+]
+
+module.exports = {
+  parse,
+  components,
+  maskable: false,
```
Comment (jgrayson676) [inline lib/types/dob.js:26]: Additional changes will be needed to support partial masking of this field, so I've left it unmasked until those changes are made.

File: test/dob.test.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      });
+
+      it('should have string value', function () {
+        const parsed = dob.parse(string);
+        assert.equal(parsed.toString(), '2014-06-02');
+        assert.equal(parsed.valueOf(), '2014-06-02');
+      });
+
+      it('should have age', function () {
+        assert.equal(dob.parse(string).age, 8);
```
Comment (josevsalina) [inline test/dob.test.js:53]: In this case, the age should be 7 instead? he has 7 years and 11 months

File: test/dob.test.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      });
+
+      it('should have string value', function () {
+        const parsed = dob.parse(string);
+        assert.equal(parsed.toString(), '2014-07-18');
+        assert.equal(parsed.valueOf(), '2014-07-18');
+      });
+
+      it('should have age', function () {
+        assert.equal(dob.parse(string).age, 7.9);
```
Comment (josevsalina) [inline test/dob.test.js:85]: and in this case is 7 years and 10 months, not sure what the .9 means

File: lib/types/dob.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+const moment = require('moment');
```
Comment (richardfordjunior) [inline lib/types/dob.js:1]: Can we use another utility, since moment is deprecated?

File: test/trustedform-url.test.js [hunk 5]
```diff
@@ -1,5 +1,6 @@
 const {assert} = require('chai');
 const moment = require('moment');
```
Comment (johnrb2) [inline test/trustedform-url.test.js:2]:
```suggestion
const {assert} = require('chai');
```
moment library is unused can be removed

File: lib/types/dob.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+const date = require('./date');
+
+const parse = function (string, req) {
+  const parsed = date.parse(string, req);
+  // parsed.masked = false;
```
Comment (johnrb2) [inline lib/types/dob.js:6]: Is this a comment for once masking is added?

File: test/dob.test.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        assert.equal(parsed.valueOf(), '2014-06-02');
+      });
+
+      it('should have age', function () {
+        assert.equal(dob.parse(string).age, 8);
+      });
+
+      it('should have year', function () {
+        assert.equal(dob.parse(string).year, 2014);
+      });
```
Comment (johnrb2) [inline test/dob.test.js:58]: How do we feel about only keeping the tests that don't have overlap with the Date type?

**Review by johnrb2 — CHANGES_REQUESTED** [review]
Great work, just a couple of changes requested and some questions.

File: test/dob.test.js (hunk 3 above)
Comment (cgrayson) [reply test/dob.test.js:85]: See above; the difference to 6/18/22 is 7 years, 11 months. And the story called for a decimal representation of the number of years. So roughly estimating by months: 7 + 11/12 = 7.9 (though the calculation is actually based on number of days.)

File: lib/types/dob.js (hunk 4 above)
Comment (cgrayson) [reply lib/types/dob.js:1]: @richardfordjunior - personally I think that kind of change is worth its own story, and moment would be replaced throughout the library all at once. Also - if there's a chosen replacement for usage in LC libraries, I don't know what it is. If not, I also think that's worth some deliberation and standardization on that. My 2¢.

**Review by cgrayson — APPROVED** [review]
Otherwise LGTM

**Review by johnrb2 — APPROVED** [review]
LGTM!

**Comment by johnrb2** [comment]
It looks like the Date test is choking on the last string in `euroStrings` array. Also not sure if we want to change L9 of `.github/workflows/test.yml` to

```yml
        node-version: [14.x, 16.x]
```

since we are on Node 14 going to 16

**Comment by johnrb2** [comment]
If you change the `lib/types/date.js` if you replace the array starting on [L9](https://github.com/activeprospect/leadconduit-types/blob/9ecae0a012d40e28abb94e32acbd4073f9d1c75c/lib/types/date.js#L9) to 

`fixes comment on L17 and adds format to L28`
```js
const formats = [
  'ddd MMM DD YYYY', // 'Mon Jun 02 2014'
  'MMM DD YYYY',     // 'Jun 02 2014'
  'M/D/YYYY',        // '6/2/2014', '06/02/2014'
  'M/D/YY',          // '6/2/14'
  'YYYY-MM-DD',      // '2014-06-02'
  'MM-DD-YYYY',      // '06-02-2014'
  'MMDDYYYY',        // '06022014'
  'YYYYMMDD',        // '20140602'

  // European formats (month first)
  'ddd DD MMM YYYY',   // 'Fri 18 July 2014'
  'DD MMM YYYY',       // '18 July 2014'
  'D/M/YYYY',          // '18/7/2014'
  'D/M/YY',            // '18/7/14'
  'YYYY-DD-MM',        // '2014-18-07'
  'DD-MM-YYYY',        // '18-07-2014'
  'DDMMYYYY',          // '18072014'
  'YYYY-DD-MM',        // '2014-18-07'
  'YYYYDDMM'           // '20141807'
];
```

**Review by johnrb2 — DISMISSED** [review]
Nice!

**Review by asedarski — DISMISSED** [review]

**Review by richardfordjunior — DISMISSED** [review]
Nice work, LGTM

**Review by cgrayson — APPROVED** [review]

**Review by robdll — APPROVED** [review]

**Pull request merged**